### PR TITLE
Story screen update

### PR DESCRIPTION
**Pull request by babaissarkar**
Modifies the story screen to give it the following look : 
Screenshot from 2024-03-06 22-41-33

**No text :**
Screenshot from 2024-03-17 19-29-58

Clicking on the left/right ornate arrow shows previous/next story part. Clicking on the "skip" text (or the dots) skips the story screen.

Addresses issue #8508, but in a different way.
**TODO**
 ~~Clicking on text moves story forward.~~ (Can't do that as a label already has an internal left click handler which deals with embedded links. Adding a listener might overwrite that.)
- [x] Skip button positioning may need to be adjusted for `text-layout=top`?
- [x] correct position of skip and left/right arrows when there's no text.

### CODE REVIEW

**Comment by babaissarkar** [comment]
44e7a4b24078947759d3af6daff13efa064b0f44 might also be able to solve https://github.com/wesnoth/wesnoth/issues/4817.

**Comment by CelticMinstrel** [comment]
I assume the escape key still skips too?

File: src/gui/dialogs/story_viewer.hpp [hunk 1]
```diff
@@ -75,6 +75,11 @@ class story_viewer : public modal_dialog
 
 	void flag_stack_as_dirty();
 
+	// Manually close the dialog, since widget with id=cancel is not a button
```
Comment (CelticMinstrel) [inline src/gui/dialogs/story_viewer.hpp:78]: Why is it not a button? Would it not be easier if it were a button?

**Comment by babaissarkar** [comment]
> I assume the escape key still skips too?

Yes.

File: src/gui/dialogs/story_viewer.hpp (hunk 1 above)
Comment (babaissarkar) [reply src/gui/dialogs/story_viewer.hpp:78]: A button would not fit with the overall design, hence the work around.

Comment (babaissarkar) [reply src/gui/dialogs/story_viewer.hpp:78]: If we want to make it a button, we need some kind of ornate button, which would also require some more assets.

Comment (CelticMinstrel) [reply src/gui/dialogs/story_viewer.hpp:81]: I think you're missing the point? A button can have any appearance you want. You can make a button that looks exactly like your skip text. (You'd probably want a different hovered appearance as well though.)

Comment (babaissarkar) [reply src/gui/dialogs/story_viewer.hpp:81]: Right, can't believe I missed that. And the hover appearance change would be a nice addition. Will try to do that.

Comment (babaissarkar) [reply src/gui/dialogs/story_viewer.hpp]: Now everything is a button (though I don't like it that canvas doesn't have relative positioning.) and it also makes the code a bit more cleaner.

**Comment by babaissarkar** [comment]
@Vultraz How can I refresh the drawing in the middle of `display_part()`? `queue_redraw()` doesn't work. It now has the small bug that if someone resizes the window in the middle of story screen, they need to click on next/prev buttons for the text to be correctly wrapped.

**Comment by babaissarkar** [comment]
The buttons now have hover and press coloring. A slightly yellow tint for hover and a deeper orange hue for press. I think it integrates neatly with wesnoth's coloring scheme, but I'm open to better color schemes.

**Comment by babaissarkar** [comment]
Added `actions` key that allows additional WFL statements to be executed to `[text]` in canvas. This functionality already existed in `[image]`.

**Comment by babaissarkar** [comment]
Looking for people who can test this on a wide monitor for #4817. (@stevecotton @Wedge009 do you have access to any such?)
@CelticMinstrel the WFL function `set_var()` doesn't seem to be documented here : https://wiki.wesnoth.org/Wesnoth_Formula_Language.

**Comment by Wedge009** [comment]
I only have access to mainly 16:9 displays (one is 16:10, at 1920x1200). I don't like 'ultra-wide' displays.

**Comment by babaissarkar** [comment]
> I only have access to mainly 16:9 displays (one is 16:10, at 1920x1200). I don't like 'ultra-wide' displays.

Don't have to be ultra-wide, just some with width > 1500 for now. I mainly want to see if the text stays correctly in the middle, and doesn't spreads too widely on the screen. I also need to know what people actually feel about this redesign. :)

**Comment by CelticMinstrel** [comment]
> @CelticMinstrel the WFL function `set_var()` doesn't seem to be documented here : https://wiki.wesnoth.org/Wesnoth_Formula_Language.

I guess it's because this rather obscure place (the `actions` key in a GUI2 canvas element) is the only place where it's currently valid, so I never got around to documenting it.

**Comment by stevecotton** [comment]
The arrows seem good on 1920x1080 and 3840x2160 screens.

I think the skip button is to prominent, if it the user is expected to use it. I suggest moving it to a corner.

The yellow of the arrows doesn't seem to match the gold of the horizontal lines, so I think the color should be shifted a bit.

**Comment by babaissarkar** [comment]
@stevecotton  How about the text? Does it get limited to the middle of the screen? Or it just hogs the entire horizontal space of the screen? (A screenshot would be helpful, preferably with a lot of text.)
The other points are good, I will think on them.

**Comment by nemaara** [comment]
Somehow the arrows don't seem to match with the rest of Wesnoth's style to me. What if they had borders and were a bit smaller? Maybe the issue is the background being a map? hmm

**Comment by babaissarkar** [comment]
> Somehow the arrows don't seem to match with the rest of Wesnoth's style to me.
It's definitely possible, since the majority of the art, like the cursor and the orbs are 3d. Would be really good if an artist can tweak them up a bit. Since we're talking about graphics, would be good to have @doofus-01 's opinion on this too.

**Comment by doofus-01** [comment]
I will look into this today

**Review by doofus-01 — CHANGES_REQUESTED** [review]
I'm not saying that the images are bad, but I agree with nemaara that the result doesn't quite fit with the rest of the UI theme.  I don't think it's a matter of shading (for psuedo-3D?), it's the curves and brightness.  Most of the other elements are angular, often with rounded or detailed corners to give a subdued fantasy theme; dark blue with gold trim, tan and wood, saying it's fantasy world, but not a grim/gothic one, nor one completely dominated by magic.

Maybe we can worry about theme/mood later.  For now just get something that works well technically, then updating the images within general size/shape constraints, and adjusting the button state IPFs, can be a separate project.

For now, this is what I would ask:
1.  Fix the jumping right-side "skip" dots.
2.  Can the L/R arrows position be held constant as the text length changes?
3.  Consider dialing back the curves and brightness/whiteness.  The tapering dots is a perfectly fine idea though.  Don't spend too much time on this point, there is plenty of time to touch-up three small images in 1.19/1.20.

This image should show some of what I meant.  The dots I added on the top two bands are a failed experiment, but they do show that the right-hand dots jumping goes away with a 69x35px image, in case that helps with troubleshooting.  
Screenshot_20240330_test

**Comment by babaissarkar** [comment]
@doofus-01 
> 1. Fix the jumping right-side "skip" dots.
Fixed.
> 2. Can the L/R arrows position be held constant as the text length changes?
Tricky, but not impossible. They are currently vertically centered on their grid cell. If I want to keep it constant, I need to calculate and then set the position manually. Though I'd like to ask @CelticMinstrel and @Vultraz 's opinions if this is a good idea from UI standpoint.
> 3. Consider dialing back the curves and brightness/whiteness. 
Current screenshot :
Screenshot from 2024-03-31 10-50-57
Screenshot from 2024-03-31 10-51-05

**Comment by babaissarkar** [comment]
Question : What's the correct place for putting the source svg file I have?

**Comment by doofus-01** [comment]
> Question : What's the correct place for putting the source svg files I have? Wesnoth currently doesn't support SVG, so can't really distribute it.

It should probably go here:  https://github.com/wesnoth/resources

**Comment by CelticMinstrel** [comment]
> They are currently vertically centered on their grid cell. If I want to keep it constant, I need to calculate and then set the position manually. Though I'd like to ask @CelticMinstrel and @Vultraz 's opinions if this is a good idea from UI standpoint.

I don't remember… does GUI2 not have an "align to bottom" option?

**Comment by babaissarkar** [comment]
> Can the L/R arrows position be held constant as the text length changes?

Fixed in commit 92e04788890520c4077c667a843fea0adaad5cbc

Also made the Skip button less prominent (comment https://github.com/wesnoth/wesnoth/pull/8510#issuecomment-2016876930) by reducing its font size and making it use the default Sans font. I also gave it a red tint, which seems to be more consistent with its purpose. (Sadly the color affects the text and the graphic differently, so the text looks more bright that I want it to be.)

**Comment by CelticMinstrel** [comment]
There seem to be some conflicts in this branch now…

**Comment by babaissarkar** [comment]
@CelticMinstrel  : I'll rebase it.

@doofus-01 : Are the changes fine, or more changes are needed?

**Comment by stevecotton** [comment]
> I didn't change [skip] to the side because that would make the design alignment-dependent (think RTL text).

Unless the arrows change meaning depending on the text direction, I don't see why it would be RTL dependent, it would just be on the same side as the "next" arrow. Swapping the arrows' meaning based on text direction might be good, but it's something that should be done consistently throughout the UI.

**Comment by babaissarkar** [comment]
@stevecotton Is this acceptable?
Screenshot from 2024-04-09 21-40-25

**Comment by stevecotton** [comment]
That looks good to me.

**Comment by Vultraz** [comment]
I’m not sure about that design. Let me see if I can come up with something.

**Comment by babaissarkar** [comment]
Here's another one :
Screenshot from 2024-04-10 08-47-57

Made the skip button a bit more prominent so that users don't confuse it with body text and know that it's actually clickable (though I have hover color change). Can get rid of the underline.

**Comment by babaissarkar** [comment]
Another nuisance is that when the map comes up, the button becomes hard to read, even with text outline on.

**Comment by babaissarkar** [comment]
This PR is essentially done, only two issues blocking it :
- [ ] In the first story screen, the text width is not being set. I've tried putting `set_max_width()` at various places, but to no avail. (@Vultraz ?)
- [ ] The buttons and the Skip text may be hard to read with no text and only the map as background. See the no text image on the original post. Suggestions on how to change the coloring would be welcome. Note that my options are limited for the text.

**Comment by babaissarkar** [comment]
Made the outline on the arrow buttons darker

**Comment by CelticMinstrel** [comment]
> * The buttons and the Skip text may be hard to read with no text and only the map as background.

For the text, the normal way to address that would be to add a font outline to enforce a consistent contrast.

**Comment by babaissarkar** [comment]
> > * The buttons and the Skip text may be hard to read with no text and only the map as background.
> 
> For the text, the normal way to address that would be to add a font outline to enforce a consistent contrast.

Yes, already using `outline = true`.

**Comment by CelticMinstrel** [comment]
…why is outline a boolean instead of a number?

That aside, it looks oddly dark. Is the text colour not set to white?

**Comment by babaissarkar** [comment]
> …why is outline a boolean instead of a number?

https://github.com/wesnoth/wesnoth/blob/39987cd1ca4c09ea2f5d9de08ecfa3c6b029fe5d/src/font/text.cpp#L557-L565

Seems to me it does not allow to specify the stroke width of the outline, only whether you want it or not.

> That aside, it looks oddly dark. Is the text colour not set to white?

I was using the default font color of "215,215,215", but I've changed it to "255,255,255" now.

**Comment by CelticMinstrel** [comment]
I think I also see your diff removing an opacity of 200? Though I'm not sure if that applies to "Skip" specifically.

**Comment by babaissarkar** [comment]
Yes, changed the opacity to full so it looks brighter.

**Review by doofus-01 — APPROVED** [review]
My concerns were addressed, thanks.

I guess there's still some cleanup that will be done for `images/misc/dots.png` and "header_story_skip"?

**Comment by babaissarkar** [comment]
This one's done, except for the first issue at https://github.com/wesnoth/wesnoth/pull/8510#issuecomment-2060642141. I've looked around, but still couldn't find what's causing it. If the user presses "next" then "previous" to come back to the first screen the width is correctly set. This simply means that #4817 cannot be closed :(
I've adjusted the text width value so that visually nothing seems out of place, however people with big screen might be able to notice it.

**Comment by ForestDragon-wesnoth** [comment]
Looks interesting, but in my opinion the old system of next/back/skip buttons being close to each other was more convenient. I think 1.18/1.16 story UI was simple but effective.

The first post's skip button may be a bit too visible, sure, but the latest one seems too obscure and hidden.

**Comment by babaissarkar** [comment]
> Looks interesting, but in my opinion the old system of next/back/skip buttons being close to each other was more convenient. I think 1.18/1.16 story UI was simple but effective.
I am afraid the entire point of the PR is to make Wesnoth more aesthetically pleasing and bring it more closer to looking like a game. The current UI and theming is effective and to the point, I don't disagree, but it looks more like an application software and not like a game.

> The first post's skip button may be a bit too visible, sure, but the latest one seems too obscure and hidden.
I agree with this one, perhaps increasing the Skip text's font size will help?

**Comment by ForestDragon-wesnoth** [comment]
> I agree with this one, perhaps increasing the Skip text's font size will help?

I think something that could help making the skip button feel like a proper button again is adding a bit of background on it, even if semi-transparent.

**Comment by babaissarkar** [comment]
> I think one of the things that could help making the skip button feel like a proper button again is adding a bit of background on it, even if semi-transparent.

Hmm, I'll think on it. Meanwhile I'm also thinking if I can bring back the ease of use the previous buttons offered without sacrificing the aesthetics.

**Comment by ForestDragon-wesnoth** [comment]
> Hmm, I'll think on it. Meanwhile I'm also thinking if I can bring back the ease of use the previous buttons offered without sacrificing the aesthetics.

A big convenient feature of the old system was that the buttons were close to eachother

**Comment by ForestDragon-wesnoth** [comment]
whether the next/previous buttons are in the new proposed style, or the 1.18-like "all three buttons next to eachother" style, I think skip should be closer to the next button maybe right under next?

**Comment by babaissarkar** [comment]
@ForestDragon-wesnoth what are your thoughts on the centered Skip button? (from the very first post)

**Comment by ForestDragon-wesnoth** [comment]
> @ForestDragon-wesnoth what are your thoughts on the centered Skip button? (from the very first post)

Looks better than the underlined non-centered text in the newer screenshot

**Comment by babaissarkar** [comment]
Reverted the previous commit and went back to the centered skip button as proposed in the original design, since @Vultraz strongly disagrees with the current Skip button design and positioning.

**Comment by grz0** [comment]
@babaissarkar I think your idea is good, but I dislike the style of the arrow images. IMO, they don't fit the rest of the game UI and they are too complex. I created a different version of the arrow image:
arrow

Could you test my image with your outro changes?

**Comment by babaissarkar** [comment]
@grz0 
Do you mean like this :
Screenshot from 2024-04-23 08-53-26

Or you want me to keep the dots and just change the arrowhead?

Note that I think any image changes should be part of a separate PR and reviewed by @doofus-01, who already agreed to this version.

**Comment by grz0** [comment]
> Do you mean like this :

Yes, thanks.

**Comment by doofus-01** [comment]
Yeah, the arrow images should probably be a separate PR.  I agree with grz0 that the originals don't fit well with the rest of the UI, but the new ones don't really seem much of an improvement.  We could be at this a while, so bettter to get the feature in front of more eyeballs, then the image can be easily updated.

**Comment by babaissarkar** [comment]
Clicking on the story text now advances the story one screen forward (dea363d8509e2442c5d29f9f5249b5a03715d36f).

**Review by Vultraz — APPROVED** [review]

**Comment by Wedge009** [comment]
> Don't have to be ultra-wide, just some with width > 1500 for now. I mainly want to see if the text stays correctly in the middle, and doesn't spreads too widely on the screen...

I didn't comment much on this because I don't have a strong opinion on the redesign. Now that it's merged, though, I did have a bit of a look through it and noticed this in the Two Brothers introduction (regarding centred text, this doesn't quite seem centred).
tb

Also, I wonder what might happen if there's an event marker that falls under the story text bar - it seems much taller than it needs to be.

**Comment by babaissarkar** [comment]
@Wedge009 
A couple of tests :
1. Can you tweak the "4" factor here to see what works for you?
https://github.com/wesnoth/wesnoth/blob/e2f893c2eaafeba1a9053576ab7ed099d1ef7110/data/gui/window/story_viewer.cfg#L23

2. Change the alignment here to "center" to see if the text gets center aligned?
https://github.com/wesnoth/wesnoth/blob/e2f893c2eaafeba1a9053576ab7ed099d1ef7110/data/gui/window/story_viewer.cfg#L74

**Comment by Vultraz** [comment]
I notice the same problem. I’ll work on it.

**Comment by babaissarkar** [comment]
@Vultraz
a couple more issues I didn't manage to deal with :
1. Happens only with my campaign, but not with any of the mainline ones :
Screenshot from 2024-05-25 07-19-27
(clicking forward and back resolves it for some reason)
2. Skip should be pegged to the bottom here :
Screenshot from 2024-05-25 07-19-43

3. Also, can you add the functionality that the story moves forward when the background image is clicked?

**Comment by Wedge009** [comment]
> Change the alignment here to "center" to see if the text gets center aligned?

image

Not as extreme as before, but still a bit to the left of actual centre. Centre is the left edge of the 'h' in 'as he could'.

> Can you tweak the "4" factor here to see what works for you?

I did a comparison with 1.18.0 and the height is the same (I didn't realise this was the case), so probably don't need to worry about this.

**Comment by Vultraz** [comment]
Issues should be fixed in 4cb9df8e6c0307e065c29d2d7aad07debdc5e0da.

**Comment by Wedge009** [comment]
From what I saw in a quick check, looks good to me.

**Pull request merged**